tochar: implement cache for parse format

otan opened the pull request:
Implemented a basic cache for parse formats, similar to `ReCache`. Haven't done any memory monitoring as the space this takes is predictable - cache size * maxCacheKeySize * some constant for vars.

Before:
```
BenchmarkTimeToChar-10        	  761224	      1577 ns/op
BenchmarkIntervalToChar-10    	  552865	      2226 ns/op
```

After:
```
BenchmarkTimeToChar-10        	 1719128	       702.8 ns/op
BenchmarkIntervalToChar-10    	 1551093	       771.4 ns/op
```

Epic: None

Release note (sql change, performance improvement): `to_char` now has caching for parse formats, marking a speedup when running to_char with the same format between sessions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91564)
<!-- Reviewable:end -->

rafiss reviewed (approved):
lgtm! but if it can be a RWMutex, let's do that. i'm good to backport

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91564#-:-NIoi9FVE7b8xS9ZRCR8:b-i5u7zv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[`pkg/util/tochar/cache.go` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91564#-NIohDB38AqNozipoxFi:-NIohDB40WYiFyOYBOZH:b-pgalgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/64a0d23668c06965c18db7d2abce1580a4538a6d/pkg/util/tochar/cache.go#L24)):*
> ```Go
> type FormatCache struct {
> 	mu struct {
> 		syncutil.Mutex
> ```

can this be a RWMutex? the cache will be mostly read heavy right?


<!-- Sent from Reviewable.io -->

otan commented:
made it RWMutex

bors r=rafiss

otan committed:
tochar: implement cache for parse format

Implemented a basic cache for parse formats, similar to `ReCache`.
Haven't done any memory monitoring as the space this takes is
predictable - cache size * maxCacheKeySize * some constant for vars.

Before:
```
BenchmarkTimeToChar-10        	  761224	      1577 ns/op
BenchmarkIntervalToChar-10    	  552865	      2226 ns/op
```

After:
```
BenchmarkTimeToChar-10        	 1719128	       702.8 ns/op
BenchmarkIntervalToChar-10    	 1551093	       771.4 ns/op
```

Release note (sql change, performance improvement): `to_char` now has
caching for parse formats, marking a speedup when running to_char with
the same format between sessions.

otan commented:
bors r=rafiss
